src/brains2/src/estimation/spline_test.cpp

```diff
@@ -0,0 +1,456 @@
+        int row = i;
+        int col = 4 * i;
+        tripletListB.emplace_back(row, col, 1.0);
+    }
+
+    // Second part of B: spkron([0, ..., 0, 1], [1, 1, 1, 1])
+    int last_row_B = N;
+    col_offset = 4 * (N - 1);
+    for (int j = 0; j < 4; ++j) {
+        tripletListB.emplace_back(last_row_B, col_offset + j, 1.0);
+    }
+
+    B.setFromTriplets(tripletListB.begin(), tripletListB.end());
+
+    // Build C
+    int C_rows = N + 1;
+    int C_cols = 4 * N;
+    Eigen::SparseMatrix<double> C(C_rows, C_cols);
+    std::vector<Eigen::Triplet<double>> tripletListC;
+
+    // First part of C
+    for (int i = 0; i < N; ++i) {
+        int row = i;
+        int col = 4 * i + 2;
+        double value = 2.0 / (delta_s[i] * delta_s[i]);
+        tripletListC.emplace_back(row, col, value);
+    }
+
+    // Second part of C: last row
+    double delta_s_last = delta_s[N - 1];
+    int last_row_C = N;
+    col_offset = 4 * (N - 1);
+    tripletListC.emplace_back(last_row_C, col_offset + 2, 2.0 / (delta_s_last * delta_s_last));
+    tripletListC.emplace_back(last_row_C, col_offset + 3, 6.0 / (delta_s_last * delta_s_last));
+
+    C.setFromTriplets(tripletListC.begin(), tripletListC.end());
+
+    // Compute P = B^T * B + curv_weight * C^T * C
+    Eigen::SparseMatrix<double> P = B.transpose() * B + curv_weight * C.transpose() * C;
+
+    // Compute q = -B^T * path
+    Eigen::VectorXd q_x = -B.transpose() * path.col(0);
+    Eigen::VectorXd q_y = -B.transpose() * path.col(1);
+
+    // Build constraint vector b
+    Eigen::VectorXd b_x(A_rows);
+    Eigen::VectorXd b_y(A_rows);
+    b_x.setZero();
+    b_y.setZero();
+    b_x[A_rows - 2] = delta_s[0] * std::cos(initial_heading);
+    b_x[A_rows - 1] = delta_s[N - 1] * std::cos(final_heading);
+
+    b_y[A_rows - 2] = delta_s[0] * std::sin(initial_heading);
+    b_y[A_rows - 1] = delta_s[N - 1] * std::sin(final_heading);
+
+    // Setup OSQP solver for X
+    OsqpEigen::Solver solver_x;
+    solver_x.settings()->setWarmStart(true);
+    solver_x.settings()->setVerbosity(verbose);
+    solver_x.data()->setNumberOfVariables(P.cols());
+    solver_x.data()->setNumberOfConstraints(A_rows);
+
+    if (!solver_x.data()->setHessianMatrix(P)) {
+        throw std::runtime_error("Failed to set P");
+    }
+    if (!solver_x.data()->setGradient(q_x)) {
+        throw std::runtime_error("Failed to set q_x");
+    }
+    if (!solver_x.data()->setLinearConstraintsMatrix(A)) {
+        throw std::runtime_error("Failed to set A");
+    }
+    solver_x.data()->setLowerBound(b_x);
+    solver_x.data()->setUpperBound(b_x);
+    if (!solver_x.initSolver()) {
+        throw std::runtime_error("Failed to initialize solver for X");
+    }
+
+    // Solve for X
+    if (solver_x.solveProblem() != OsqpEigen::ErrorExitFlag::NoError) {
+        throw std::runtime_error("Failed to solve QP for X");
+    }
+
+    Eigen::VectorXd p_X = solver_x.getSolution();
+
+    // Setup OSQP solver for Y
+    OsqpEigen::Solver solver_y;
```

Comment:
Reuse the same solver as for X (avoids reallocation of memory) and only override q and b